runtime/Java/src/org/antlr/v4/runtime/tree/TreeRewriteSupport.java

```diff
@@ -0,0 +1,157 @@
+package org.antlr.v4.runtime.tree;
+
+
+import org.antlr.v4.runtime.ParserRuleContext;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/** This class encapsulates code necessary to support tree rewriting
+ *  using visitors. To rewrite trees using ANTLR-generated visitors,
+ *  override two key methods:
+ * <pre>
+	 public class MyTreeRewriter extends MyParserNameBaseVisitor<ParseTree> {
```

Comment:
Yeah, originally I was going to just generate a new base class called XRewriteVisitor and thereby avoid all this overriding of visitChildren etc... but it means generating a new file, a big burden on all target maintainers and a biggish change to documentation.

Intrigued by BiFunction idea. How would it do matching?  By root classname? user has to do the matching in BiFunction?

TreeEditor sounds interesting too for common replacements.

Re `null` return.  I supply a set/delete child but detecting when to delete could require significant parsing of the child, which means visiting functionality.  Perhaps not `null` but a special `DEADTREE`?